R/html_notebook.R

```diff
@@ -0,0 +1,204 @@
+#'
+#' Parse an HTML notebook, retrieving annotation information
+#' related to generated outputs in the document, as well as the
+#' original R Markdown source document.
+#'
+#' @param path The path to an R Notebook file (with extension \code{.nb.html}).
+#' @param encoding The document's encoding (assumend \code{"UTF-8"} by default).
+#' @export
+parse_notebook <- function(path, encoding = "UTF-8") {
+
+  contents <- read_lines_utf8(path, encoding = encoding)
+
+  re_comment  <- "^\\s*<!--\\s*rnb-([^-]+)-(begin|end)\\s*([^\\s-]+)?\\s*-->\\s*$"
+  re_document <- "^\\s*<!--\\s*rnb-document-source\\s*([^\\s-]+)\\s*-->\\s*$"
+
+  rmd_contents <- NULL
+  builder <- list_builder()
+
+  for (row in seq_along(contents)) {
+    line <- contents[[row]]
+
+    # extract document contents
+    matches <- gregexpr(re_document, line, perl = TRUE)[[1]]
+    if (!identical(c(matches), -1L)) {
+      start <- c(attr(matches, "capture.start"))
+      end   <- start + c(attr(matches, "capture.length")) - 1
+      decoded <- rawToChar(base64enc::base64decode(substring(line, start, end)))
+      rmd_contents <- strsplit(decoded, "\\r?\\n", perl = TRUE)[[1]]
+      next
+    }
+
+    # extract information from comment
+    matches <- gregexpr(re_comment, line, perl = TRUE)[[1]]
+    if (identical(c(matches), -1L))
+      next
+
+    starts <- c(attr(matches, "capture.start"))
+    ends   <- starts + c(attr(matches, "capture.length")) - 1
+    strings <- substring(line, starts, ends)
+
+    n <- length(strings)
+    if (n < 2)
+      stop("invalid rnb comment")
+
+    # decode comment information and update stack
+    data <- list(row = row,
+                 label = strings[[1]],
+                 state = strings[[2]])
+
+    # add metadata if available
+    if (n >= 3 && nzchar(strings[[3]]))
+      data[["meta"]] <- base64_decode_object(strings[[3]])
+    else
+      data["meta"] <- list(NULL)
+
+    # append
+    builder$append(data)
+  }
+
+  annotations <- builder$data()
+
+  # extract header content
+  head_start <- grep("^\\s*<head>\\s*$",  contents, perl = TRUE)[[1]]
+  head_end   <- grep("^\\s*</head>\\s*$", contents, perl = TRUE)[[1]]
+
+  list(source = contents,
+       rmd = rmd_contents,
+       header = contents[head_start:head_end],
+       annotations = annotations)
+}
+
+notebook_annotated_output <- function(output, label, meta = NULL) {
+  before <- if (is.null(meta)) {
+    sprintf("\n<!-- rnb-%s-begin -->\n", label)
+  } else {
+    meta <- base64_encode_object(meta)
+    sprintf("\n<!-- rnb-%s-begin %s -->\n", label, meta)
+  }
+  after  <- sprintf("\n<!-- rnb-%s-end -->\n", label)
+  paste(before, output, after, sep = "\n")
+}
+
+notebook_annotated_knitr_hook <- function(label, hook, meta = NULL) {
+  force(list(label, hook, meta))
+  function(x, ...) {
+    output <- hook(x, ...)
+    meta <- if (is.function(meta)) meta(x, output, ...)
+    notebook_annotated_output(output, label, meta)
+  }
+}
+
+knitr_options_notebook <- function() {
```

Comment:
Okay, let's discuss after you've thought it through some more. If you absolutely need to make it all pluggable I'm okay with that (but we may want to have the default be `NULL` rather than calling a function so that we don't need to export that function as well.